src/core/socialprovider.js

```diff
@@ -477,12 +478,64 @@ XMPPSocialProvider.prototype.onOnline = function(continuation) {
   this.vCardStore.updateProperty(this.id, 'status', 'ONLINE');
   this.vCardStore.refreshContact(this.id, null);
 
+  this.pollForDisconnect_();
+
   continuation(this.vCardStore.getClient(this.id));
 };
 
+XMPPSocialProvider.prototype.ping_ = function() {
+  var pingDate = new Date();
+  var ping = new window.XMPP.Element('iq', {type: 'get'})
+      .c('ping', {'xmlns': 'urn:xmpp:ping'}).up();
+  this.client.send(ping);
+
+  // Check that we got a response from the server after the ping was sent.
+  setTimeout(function() {
+    if (this.client &&  // check this.client to be sure logout wasn't called.
+        (!this.lastMessageDate_ || this.lastMessageDate_ < pingDate)) {
+      // No response to ping, we are disconnected.
+      console.warn('No ping response from server, logging out');
+      this.logout();
+    }
+  }.bind(this), 10000);
+};
+
+XMPPSocialProvider.prototype.pollForDisconnect_ = function() {
+  if (this.pollForDisconnectInterval_) {
+    console.error('pollForDisconnect_ called while already polling');
+    return;
+  }
+
+  var MAX_MS_WITHOUT_COMMUNICATION = 60000;
+  var lastAwakeDate = new Date();
+  this.pollForDisconnectInterval_ = setInterval(function() {
+    // Check if the computer had gone to sleep
+    var nowDate = new Date();
+    if (nowDate - lastAwakeDate > 2000) {
+      // Timeout expected to run every 1000 ms didn't run for over 2000 ms,
+      // probably because the computer went to sleep.  Send a ping to check
+      // that we are still connected to the XMPP server.
+      console.log('Detected sleep for ' + (nowDate - lastAwakeDate) + 'ms');
+      this.ping_();
+    }
+    lastAwakeDate = nowDate;
+
+    // Check that we are still receiving data from the XMPP server, about
+    // once per minute (randomized).
+    if (nowDate.getSeconds() === Math.floor(Math.random() * 60) &&
```

Comment:
why not `if (nowDate.getSeconds() === 0 && ...`.  No reason to make it nondeterministic